cupy/math/misc.py

```diff
@@ -63,15 +66,39 @@ def _fft_convolve(a1, a2, mode):
 
     if a1.dtype.kind == 'c' or a2.dtype.kind == 'c':
         fft, ifft = cupy.fft.fft, cupy.fft.ifft
+        is_c2c = True
     else:
         fft, ifft = cupy.fft.rfft, cupy.fft.irfft
+        is_c2c = False
+
+    out_dtype = cupy.result_type(a1, a2)
+
+    # hack to work around NumPy/CuPy FFT dtype incompatibility
+    if out_dtype == cupy.float16:
+        dtype = cupy.float32
```

Comment:
We don't really want to compute a fp16 fft in fp32 if possible 😅

Reply:
What do you mean? cuFFT does not support fp16, and we internally cast fp16 to fp32 before executing FFT. This is just to do the cast step earlier, and also to generate a correct plan.

Reply:
Ok! I was not understanding it completely so it's my fault :).

I verified what you said. Could you please add a comment detailing this?
Cufft does not support fp16 so we cast it to fp32 here instead of letting cupy.fft do it to create a correct plan.

Reply:
Thanks! I added it. I also call the internal fft helper to determine the type. Seems less error prone to me.